operator creates the Service Monitor

laurafitzgerald opened the pull request:
**JIRA**
https://issues.jboss.org/browse/AEROGEAR-9989

**Note**
Work in progress at the moment, sharing for early feedback - items below are in order of commits pushed.
- [x] ServiceMonitor
- [x] Formatting and Version improvements
- [x] Prometheus Rule
- [x] Grafana Dashboard
- [x] Fix the Grafana Dashboard
- [x] Check that the ServiceMonitor CRD exist in the cluster
- [x] Check that the Service PrometheusRule CRD exist in the cluster
- [x] Check that the GrafanaDashboard CRD exist in the cluster
- [x] Trigger the Reconcile loop using SyncPeriod


**IMAGE**
quay.io/aerogear/mobile-developer-console-operator:2aaadfb

**Verification**
_Steps_

**Verify functionality when monitoring resources ARE NOT present on the cluster**

Ensure you've removed the monitoring resources using hte following commands:
`oc delete -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/example/prometheus-operator-crd/servicemonitor.crd.yaml`

`oc delete -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/example/prometheus-operator-crd/prometheusrule.crd.yaml`

`oc delete -f https://raw.githubusercontent.com/integr8ly/grafana-operator/master/deploy/crds/GrafanaDashboard.yaml`

Install the operator:

Update the image in operator.yaml
make cluster/clean
make cluster/prepare
make install-operator
make install-mdc

Verify that the operator didn't try to create any of the monitoring resources as it checks whether they are present on the cluster before adding watches/trying to create any crs and that it doesn't crash or error. Info logs are fine.

**Verify functionality when monitoring resources ARE present on the cluster**

`oc create -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/example/prometheus-operator-crd/servicemonitor.crd.yaml`

`oc create -f https://raw.githubusercontent.com/coreos/prometheus-operator/master/example/prometheus-operator-crd/prometheusrule.crd.yaml`

`oc create -f https://raw.githubusercontent.com/integr8ly/grafana-operator/master/deploy/crds/GrafanaDashboard.yaml`

_Expected Results_
1) Watches are setup by the operator, the logs should include the following:
```

{"level":"info","ts":1570809072.4111557,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"mobiledeveloperconsole-controller","source":"kind source: /, Kind="}
--
  | {"level":"info","ts":1570809073.4126387,"logger":"controller_mobiledeveloperconsole","msg":"Watch created for 'ServiceMonitor' resource in 'mobiledeveloperconsole-controller'"}
  | {"level":"info","ts":1570809077.1130168,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"mobiledeveloperconsole-controller","source":"kind source: /, Kind="}
  | {"level":"info","ts":1570809077.5143685,"logger":"controller_mobiledeveloperconsole","msg":"Watch created for 'PrometheusRule' resource in 'mobiledeveloperconsole-controller'"}
  | {"level":"info","ts":1570809077.7105849,"logger":"kubebuilder.controller","msg":"Starting EventSource","controller":"mobiledeveloperconsole-controller","source":"kind source: /, Kind="}
  | {"level":"info","ts":1570809078.416391,"logger":"controller_mobiledeveloperconsole","msg":"Watch created for 'GrafanaDashboard' resource in 'mobiledeveloperconsole-controller'"}


```

The reconcile will trigger within 2 minutes and the crs are then created - you'll need to refresh your browser to load the new CRDs to the Other Resources Dropdown

2) A Service Monitor is created which looks like this:

![Screenshot 2019-09-30 at 12 40 52](https://user-images.githubusercontent.com/6498727/65875647-9748e780-e37f-11e9-80c0-8a7f0473ad65.png)

3) A Prometheus Rule is created which looks like this:

![Screenshot 2019-09-30 at 15 39 42](https://user-images.githubusercontent.com/6498727/65889097-907a9e80-e398-11e9-9649-da618e36643a.png)

4) A Grafana Dashboard is created which looks like this:

![Screenshot 2019-10-01 at 11 54 54](https://user-images.githubusercontent.com/6498727/65956459-59f95e00-e442-11e9-8a1a-a40af3b18b07.png)

laurafitzgerald committed:
operator creates the Service Monitor

laurafitzgerald committed:
version and comment improvements

laurafitzgerald committed:
operator creates mdc Prometheus Rule

laurafitzgerald committed:
operator creates mdc Grafana Dashboard

laurafitzgerald committed:
add namespace to make command

laurafitzgerald committed:
check service monitor crd is present on cluster

laurafitzgerald commented:
@aliok @grdryn seeing as there will be a new approach provided by @davidffrench tomorrow 
around solving the issue of needing to stop and start the operator to add the Watches, wdyt about verifying the functionality of the operator adding the resources with a prereq added that the CRDs must be on the cluster and getting this one merged so that there isn't too many changes here.

grdryn commented:
@laurafitzgerald sounds good to me :+1:

davidffrench commented:
My PR is now up https://github.com/keycloak/keycloak-operator/pull/40

laurafitzgerald commented:
Thanks @davidffrench I didn't have time to work on this one today. I'll pick this up in the morning and add the additional logic for adding the watch when a CRD is created on the cluster later. 👍

laurafitzgerald committed:
update grafanadashboard based on INTLY-3315

laurafitzgerald committed:
create the watch and cr when a crd is created after deployment for Service Monitor

laurafitzgerald commented:
Checking why build is failing....

laurafitzgerald committed:
add line at end of file

laurafitzgerald committed:
check for crds before creating crs

laurafitzgerald committed:
trigger reconcile every 2 minutes

laurafitzgerald commented:
@aliok @grdryn @jomrazek this pr is ready for Verification now.

aliok commented:
:eye:

laurafitzgerald committed:
remove additional links

laurafitzgerald commented:
I'm also covering this JIRA: https://issues.jboss.org/browse/AEROGEAR-10026 in this pr which is a small change. It's pushed in the last commit.

aliok commented:
@laurafitzgerald I verified the steps you described above. All good.

However, I didn't really check if these serviceMonitors/rules/GrafanaDashboards look good on actual Prometheus/Grafana instances.

Do we care at this moment?

laurafitzgerald commented:
@aliok I didn't verify those either. The resources created are copies of the ones created previously and would have been verified in the last release. I plan to do a deployment on pds to verify [this pr](https://github.com/integr8ly/installation/pull/1054) I think we can do any alert/dashboard verification there.

aliok commented:
Ok then, giving my approval then

grdryn commented:
Should any of the new dependencies in Gopkg.lock be added to Gopkg.toml?

grdryn commented on cmd/manager/main.go:
```diff
@@ -105,18 +110,20 @@ func main() {
 		os.Exit(1)
 	}
 
+	syncperiod := time.Duration(time.Minute)
```
Is a minute a common sync period? It seems very low in comparison to the default of 10 hours :)

laurafitzgerald replied:
Yeah, I'm open to discussion on this time. If the grafanadashboard, servicemonitor and prometheusrule CRDs are created as part of an installation of application monitoring, how long should the user wait in order for the related crs to be created inside the reconcile?

there's also the argument that for application-monitoring it will already be installed by the time this operator is installed. We could not trigger it on a syncperiod and document it instead?

grdryn replied:
> [...] how long should the user wait in order for the related crs to be created inside the reconcile?

Not trying to give a silly answer, but they should wait up to as long as whatever we set the sync period here :)

The reason I say that (as silly as it may sound), is that we shouldn't be setting this just based on these monitoring resources, since this will reconcile everything (all primary and secondary resources). Granted for this operator we've only got 1 CRD and only allow 1 CR for that, so in this case it won't be a huge amount of extra work if this is a short period. Think about the UPS operator though: it could have a lot of CRs to manage, which could cause a lot of extra unnecessary traffic to the api server if it's set too low.

I honestly don't know what a good period is here, I'm just being cautious. Maybe a minute is fine :) @davidffrench @maleck13 do you have any thoughts on it?

grdryn commented on cmd/manager/main.go:
```diff
@@ -148,8 +155,28 @@ func main() {
+		os.Exit(1)
+	}
+
+	// Setup Scheme for Integreatly apis
```
Would it be useful to mention that this is for Grafana?

laurafitzgerald replied:
Yes, I think it would. I'll update it.

grdryn commented on cmd/manager/main.go:
```diff
@@ -190,6 +217,9 @@ func main() {
 				log.Info("Install prometheus-operator in you cluster to create ServiceMonitor objects", "error", err.Error())
 			}
 		}
+		// If this operator is deployed to a cluster without the prometheus-operator running, it will return
```
This comment seems out of place, should it be _above_ something, or is it intentionally here?

laurafitzgerald replied:
Yes, I think this was something to do with a rebase. I'll fix it up.

laurafitzgerald commented:
@grdryn I'm not sure. Is that a thing we need to do manualy?

grdryn commented on pkg/controller/mobiledeveloperconsole/mobiledeveloperconsole.go:
```diff
@@ -344,3 +347,960 @@ func newMobileClientAdminRole(cr *mdcv1alpha1.MobileDeveloperConsole) (*rbacv1.R
+	labels := map[string]string{
+		"monitoring-key": "middleware",
+		"prometheus":     "application-monitoring",
+		"role":           "alert-rules",
```
Out of curiosity, do you know if these labels are used by anything?

laurafitzgerald replied:
The variable is being used the Prometheusrule object being created a few lines below. If the question is if they are being used for something in OpenShift when it's actually deployed I need to check it. I pinged in the integreatly channel about it.

laurafitzgerald committed:
bump version

grdryn commented on pkg/controller/mobiledeveloperconsole/mobiledeveloperconsole.go:
```diff
@@ -344,3 +347,960 @@ func newMobileClientAdminRole(cr *mdcv1alpha1.MobileDeveloperConsole) (*rbacv1.R
+	}
+	objectMetaName := util.ObjectMeta(&cr.ObjectMeta, "mdc").Name
+	container := "mdc"
+	namespace := cr.Namespace
```
Is this useful? 3 lines below it, `cr.Namespace` is referenced directly, so should that line be updated or should this line be removed?

laurafitzgerald replied:
it's used in the json block below. but yes it can be used 3 lines below. Will fix.

grdryn commented on pkg/controller/mobiledeveloperconsole/mobiledeveloperconsole_controller.go:
```diff
@@ -2,23 +2,36 @@ package mobiledeveloperconsole
+
 	mdcv1alpha1 "github.com/aerogear/mobile-developer-console-operator/pkg/apis/mdc/v1alpha1"
 	"github.com/aerogear/mobile-developer-console-operator/pkg/config"
+	integreatlyv1alpha1 "github.com/integr8ly/grafana-operator/pkg/apis/integreatly/v1alpha1"
```
I wasn't going to comment on it, but in some of the other places where this package is imported it's called `integreatlyv1`, where this is more correctly called `integreatlyv1alpha1`.

grdryn commented on pkg/controller/mobiledeveloperconsole/mobiledeveloperconsole_controller.go:
```diff
@@ -27,32 +40,46 @@ var (
 	}
 
-	// Watch for changes to primary resource MobileDeveloperConsole
+	// Watch for deletion of primary resource MobileDeveloperConsole
```
Is the update to add `deletion of` accurate in these comments? AIUI this is what notices _creation_ of the MDC CR also, and also triggers on updates to it (similarly, the watches for the secondary resources will trigger on update to those resources, not just deletion).

laurafitzgerald replied:
Yes, I think you are right for the MDC cr. I'd need to test it but for the secondary resources those can be updated and the operator wont do anything in that case. Is it only this one your comment is about or is it all of the comments?

grdryn replied:
It's about all the comments -- regardless of what we do when the reconcile method is triggered, these watches will (correctly) trigger it for any changes.

grdryn commented on pkg/controller/mobiledeveloperconsole/mobiledeveloperconsole_controller.go:
```diff
@@ -115,6 +142,25 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+		for gvk := range autodetectChannel {
+			// Check if this channel event was for the PrometheusRule resource type
+			if gvk.String() == monitoringv1.SchemeGroupVersion.WithKind(monitoringv1.PrometheusRuleKind).String() {
+				watchSecondaryResource(c, gvk, &monitoringv1.PrometheusRule{}) // nolint
```
What's the `// nolint` about?

laurafitzgerald replied:
copy pasta. will remove.

grdryn reviewed (commented):
I think the resources being created could have more consistent naming, wdyt?

Also (not a fault with this PR, just a grumble comment), it would have been nice if the grafana-operator went with `grafana.integreatly.org` rather than `integreatly.org` for their group.

laurafitzgerald committed:
updates based on feedback

laurafitzgerald commented:
@grdryn I've addressed all feedback items bar the sync period as that's still under discussion.

What do you mean by `I think the resources being created could have more consistent naming, wdyt?`

grdryn commented on version/version.go:
```diff
@@ -1,5 +1,5 @@
 
 var (
-	Version = "0.2.6"
+	Version = "0.3.0"
```
If you're going to tag this once merged, should the CHANGELOG be updated as part of this also?

Also feel free to set the image tag in operator.yaml to the release version also, rather than master.

grdryn reviewed (dismissed):
What I meant about the resources having inconsistent naming is:

- ServiceMonitor: mobile-developer-console
- PrometheusRule: mdc-monitoring
- GrafanaDashboard: mdc-application

I think at least they should have the same name, and I also think it's good practice for secondary resources created by an operator to have the name of the CR that caused their creation. So these could be called e.g. `cr.Namespace + "-mdc"`.

I realise that they've probably inherited their current names from some bad decision that we made in the integreatly installer at some point, but now is definitely an opportunity to re-think that. What do you think?

laurafitzgerald committed:
version and changelog
